src/idatetime.c

```diff
@@ -143,6 +145,29 @@ SEXP convertDate(SEXP x, SEXP type)
     else if (!strcmp(ctype_str, "yearqtr")) { ctype = YEARQTR; ansint = false; }
     else internal_error(__func__, "invalid type, should have been caught before"); // # nocov
     
+    if (ctype == WEEK) {
+        if (!week_warning_issued) {
+            SEXP dt_week_warn_opt = GetOption(install("datatable.warn.week.change"), R_NilValue);
+            if (!isLogical(dt_week_warn_opt) || LOGICAL(dt_week_warn_opt)[0] != FALSE) {
+                for (int i = 0; i < n; i++) {
+                    if (ix[i] == NA_INTEGER) continue;
+
+                    int yday;
+                    convertSingleDate(ix[i], YDAY, &yday);
+
+                    int old_week = (yday / 7) + 1;
+                    int new_week = ((yday - 1) / 7) + 1;
+
+                    if (new_week != old_week) {
+                        Rf_warning("data.table::week() behavior has changed and is now sequential (day 1-7 is week 1). The first week of the year may differ from previous versions. To suppress this warning, run: options(datatable.warn.week.change = FALSE)");
+                        week_warning_issued = 1;
+                        break;
+                    }
+                }
+            }
+        }
+    }
```

Comment:
To avoid iterating over `x` and calling `convertSingleDate` twice I would expect to fully branch the `ctype==week` out. I know this will lead to duplicate code, but we will also phase this code out again after some releases.